src/utils/exportUtils/HtmlExport.ts

```diff
@@ -0,0 +1,459 @@
+            </div>`: ``}
+        ${isReply ?
+        `   <div class="mx_reply_to mx_details">
+                In reply to <a href="#${replyId}">this message</a>
+            </div>`: ``}
+        ${messageBody}
+        </div>
+    </div>
+    `;
+};
+
+
+const createHTML = async (events: MatrixEvent[], room: Room) => {
+    let content = "";
+    let prevEvent = null;
+    for (const event of events) {
+        // As the getContent of the edited event fetches the latest edit, there is no need to process edit events
+        if (isEdit(event)) continue;
+        content += dateSeparator(event, prevEvent);
+
+        if (event.getType() === "m.room.message") {
+            const replyTo = getBaseEventId(event);
+            const shouldBeJoined = prevEvent && prevEvent.getContent().msgtype === "m.text"
+            && event.sender.userId === prevEvent.sender.userId && !dateSeparator(event, prevEvent) && !replyTo;
+
+            const body = await createMessageBody(event, shouldBeJoined, !!replyTo, replyTo);
+            content += body;
+        } else {
+            const eventText = textForEvent(event);
+            content += eventText ? `
+            <div class="mx_message mx_service" id="${event.getId()}">
+                <div class="mx_body mx_details">
+                    ${textForEvent(event)}
+                </div>
+            </div>
+            ` : "";
+        }
+        prevEvent = event;
+    }
+    return wrapHTML(content, room);
+};
+
+const avatars = new Map();
+let zip: any;
+
+const exportAsHTML = async (res: MatrixEvent[], room: Room) => {
+    zip = new JSZip();
+
+    const html = await createHTML(res, room);
+
+    zip.file("index.html", html);
+    zip.file("css/style.css", css);
+
+    avatars.clear();
+
+    const filename = `matrix-export-${new Date().toISOString()}.zip`;
+
+    //Generate the zip file asynchronously
+    const blob = await zip.generateAsync({ type: "blob" });
+
+    //Create a writable stream to the directory
+    const fileStream = streamSaver.createWriteStream(filename, blob.size);
+    const writer = fileStream.getWriter();
+
+    // Here we chunk the blob into pieces of 10 MB, the size might be dynamically generated.
+    // This can be used to keep track of the progress
+    const sliceSize = 10 * 1e6;
+    for (let fPointer = 0; fPointer < blob.size; fPointer += sliceSize) {
+        const blobPiece = blob.slice(fPointer, fPointer + sliceSize);
+        const reader = new FileReader();
+
+        const waiter = new Promise<void>((resolve, reject) => {
+            reader.onloadend = evt => {
+                const arrayBufferNew: any = evt.target.result;
+                const uint8ArrayNew = new Uint8Array(arrayBufferNew);
+                writer.write(uint8ArrayNew);
+                resolve();
+            };
+            reader.readAsArrayBuffer(blobPiece);
+        });
+        await waiter;
+    }
+    writer.close();
+}
+
+export default exportAsHTML;
```

Comment:
would be good to wrap these methods in a class for simpler extensions, say someone wanted to change how date separators worked, they would just extend one method and voila

Reply:
Sorry, I don't really understand what you mean. Could you give me an example?

Reply:
Instead of having a bunch of loose methods you would make a class which implements an interface/abstract class you define.

E.g

```typescript

abstract class Exporter {
// define method signatures here
}

class HTMLExporer extends Exporter {
// define your loose functions as methods in this class e.g.:
protected async createMessageBody() {...}
protected async createHTML() {...}
public async export() {...}
}
```

Reply:
Makes perfect sense. Thank you!